components/script/dom/elementinternals.rs

```diff
@@ -0,0 +1,371 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+
+use dom_struct::dom_struct;
+use html5ever::local_name;
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::ElementInternalsBinding::{
+    ElementInternalsMethods, ValidityStateFlags,
+};
+use crate::dom::bindings::codegen::UnionTypes::FileOrUSVStringOrFormData;
+use crate::dom::bindings::error::{Error, ErrorResult, Fallible};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::{reflect_dom_object, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot, MutNullableDom};
+use crate::dom::bindings::str::{DOMString, USVString};
+use crate::dom::element::Element;
+use crate::dom::file::File;
+use crate::dom::htmlelement::HTMLElement;
+use crate::dom::htmlformelement::{FormDatum, FormDatumValue, HTMLFormElement};
+use crate::dom::node::{window_from_node, Node};
+use crate::dom::nodelist::NodeList;
+use crate::dom::validation::{is_barred_by_datalist_ancestor, Validatable};
+use crate::dom::validitystate::{ValidationFlags, ValidityState};
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+enum SubmissionValue {
+    File(DomRoot<File>),
+    FormData(Vec<FormDatum>),
+    USVString(USVString),
+    None,
+}
+
+#[dom_struct]
+pub struct ElementInternals {
+    reflector_: Reflector,
+    // if attached is false, we're using this to hold form-related state
+    // on an element for which attachInternals() wasn't called yet; this is
+    // necessary because it might have a form owner.
```

Comment:
```suggestion
    /// If `attached` is false, we're using this to hold form-related state
    /// on an element for which `attachInternals()` wasn't called yet; this is
    /// necessary because it might have a form owner.
```